### PR TITLE
Merge Rest api with knowledge gaps into master

### PR DESCRIPTION
**Pull request by JackJackie**

### CODE REVIEW

File: smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java
Comment (bnouwt) [inline smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java:81]: Two thoughts about the `KnowledgeGapSet` and `KnowledgeGap` classes:
- I would not have created the separate `KnowledgeGapSet` class, because that is not that much of a difference from `Set<KnowledgeGap>` and it is also not very common in Java to do it like that.
- I am in doubt about whether to create the `KnowledgeGap` class, because on the one hand they are `just` graph patterns and we represent those in the reasoner as `Set<TriplePattern>`. Other the other hand, they are not entirely graph patterns because they are typically more spread out over the reasoning graph.

Suggestion: remove the `KnowledgeGapSet` class and keep the `KnowledgeGap` class.

File: smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java
Comment (bnouwt) [inline smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java:191]: We might consider making the inclusion of knowledge gaps configurable, because I can imagine not all scenarios are interested in those and they might add some performance penalty. We can, however, also do that at a later stage when improving the configuration of the knowledge engine (#512) is done.

File: smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/AskResult.java
Comment (bnouwt) [inline smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/AskResult.java:112]: This addition currently does not do much, since the `KnowledgeGapSet` has no custom `toString()` method, so it just shows something like `KnowledgeGapSet@23423423`. But, if we decide to remove this class altogether (see comment below) then this is of course not necessary 👍

File: smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/KnowledgeGap.java
Comment (bnouwt) [inline smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/KnowledgeGap.java:26]: Consider using the `this.addAll(...)` method.

File: smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/KnowledgeGapSet.java
Comment (bnouwt) [inline smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/KnowledgeGapSet.java:15]: _If_ we decide to remove this class, we should move these comments to some other location. Maybe the `KnowledgeGap` class or the `getKnowledgeGaps(...)` method.

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (bnouwt) [inline smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:53]: Can we, instead of a separate REST operation for the knowledge gaps, include the knowledge gaps in the `/ask` operation when they are available? And leave them out altogether when they are not available?

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (bnouwt) [inline smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:148]: I would expect the knowledge gaps to be included in the `AskResult` here to be returned to the user of the REST API. Or do you return those in some other way?

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (bnouwt) [inline smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:180]: Before merging, these should be made `debug` again, otherwise the logs output get too cluttered.

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (bnouwt) [inline smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:256]: Return to `debug`, see comment above.

File: smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java
Comment (bnouwt) [inline smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java:740]: Consider moving most of these `info` level log statements to `debug` level.

**Review by bnouwt — CHANGES_REQUESTED** [review]
Hey @JackJackie, I finished the first review. What do you think?

File: smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java
Comment (JackJackie) [reply smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java:81]: @bnouwt : Ok, I created the `KnowledgeGapSet` because we are also using the `BindingSet` in the Smart Connector API. But, I agree that the added value of a separate `KnowledgeGapSet` class is not very high, except for the readability/clarity of the code.

Concluding: I will remove the `KnowledgeGapSet` class and replace with `Set<KnowledgeGap>`

File: smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java
Comment (JackJackie) [reply smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java:191]: Ok...for now, the calculation of knowledge gaps is only executed when there is an empty binding. This seems to be THE situation when this needs to happen. But, of course it might be better to let the user/caller decide whether or not to calculate knowledge gaps.

File: smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/AskResult.java
Comment (JackJackie) [reply smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/AskResult.java:112]: Well, it does add information in the AskResult class which is being logged and shown as INFO at the `ProactiveApiServiceImpl` class upon receipt of the AskResult.

File: smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/KnowledgeGap.java
Comment (JackJackie) [reply smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/KnowledgeGap.java:26]: Done!

File: smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/KnowledgeGapSet.java
Comment (JackJackie) [reply smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/KnowledgeGapSet.java:15]: Done!

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (JackJackie) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:53]: Yes, but this is work in progress....the current exercise was to have the knowledge gaps being nicely provided via the java API.

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (JackJackie) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:180]: Done!

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (JackJackie) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:256]: Done!

File: smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java
Comment (JackJackie) [reply smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java:740]: Done!

File: smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/AskResult.java
Comment (bnouwt) [reply smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/AskResult.java:112]: Ok, maybe I didn't look correctly, but I expected it to just show something like `KnowledgeGapSet@23423423` after the `knowledgeGaps=`. But if the `KnowledgeGapSet` is replaced by `Set<KnowledgeGap>` the `toString(...)` is more informative.

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (JackJackie) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:148]: This is my next step...

**Comment by JackJackie** [comment]
I'll continue with the REST API changes and will produce a new pull request after that

**Comment by bnouwt** [comment]
I think you can just keep working in this pull request.

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (bnouwt) [inline smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:132]: It is a bit confusing that it currently shows so many (green) changes in this file, while in reality not *that* much has changed. Maybe this is caused by the old `scAskOldPost(...)` still being there?

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (bnouwt) [inline smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:179]: We need to make sure the `tp.toString()` method actually returns valid SPARQL 1.1 triple patterns (that can be processed further if necessary), because currently [it doesn't](https://github.com/TNO/knowledge-engine/blob/rest_api_with_knowledge_gaps/reasoner/src/main/java/eu/knowledge/engine/reasoner/api/TriplePattern.java#L94-L96). Typically, `toString()` is only meant for development purposes, so maybe we could do something like this here:

```
String tpString = FmtUtils.stringForTriple(tp.asTriple(), new PrefixMappingZero());
listKnowledgeGap.add(tpString);
```

This uses the Apache Jena internal `FmtUtils.stringForTriple(...)` method for creating a representation of a triple that follows the official SPARQL 1.1 specification.

This does, however, require an additional method in the [`TriplePattern`](https://github.com/TNO/knowledge-engine/blob/rest_api_with_knowledge_gaps/reasoner/src/main/java/eu/knowledge/engine/reasoner/api/TriplePattern.java) class:

```
public Triple asTriple() {
   return Triple.create(this.getSubject(), this.getPredicate(), this.getObject());
}
```

File: smart-connector-rest-server/src/main/resources/openapi-sc.yaml
Comment (bnouwt) [inline smart-connector-rest-server/src/main/resources/openapi-sc.yaml:681]: Should we make the knowledge gaps optional? I.e. remove them from the required attribute list? In some scenarios (like knowledge gaps configured to be turned off or there not being any knowledge gaps) it might not be required to have the (possibly empty) knowledge gaps list. Not sure, though 😄

File: smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java
Comment (bnouwt) [inline smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java:612]: I don't fully understand my own comments above 😏 . I think we should just remove this whole paragraph, because it is explained in the `@return ...` part below quite clearly👍 . What do you think?

**Review by bnouwt — DISMISSED** [review]
Hey @JackJackie, looks good 👍 

If we make sure that the `TriplePattern`s are represented correctly and we have a unit test (either via the Java API (easy) or via the REST API (difficult)) and all the tests are still succeeding, we are ready to go!

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (JackJackie) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:132]: true, I will delete the scAskOldPost

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (JackJackie) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java]: Agree...actually, this was a "bug" in the sense that the returned knowledge gaps were not correct! I made the proposed changes

File: smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java
Comment (JackJackie) [reply smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java:612]: Agreed and done

File: smart-connector-rest-server/src/main/resources/openapi-sc.yaml
Comment (JackJackie) [reply smart-connector-rest-server/src/main/resources/openapi-sc.yaml:681]: @bnouwt : I think it would be good when the Rest server Ask request gets a boolean flag named `KnowledgeGapsEnabled` that indicates whether or not the user wants to get knowledge gap information. Similar to the `reasonerEnabled` flag when adding a KB. Agreed?

Comment (bnouwt) [reply smart-connector-rest-server/src/main/resources/openapi-sc.yaml:681]: Not sure yet what we should go for.

I am also wondering whether this configuration option should be configured at Knowledge Interaction (KI) *activation* (so when calling the `POST /sc/ask` operation) or at KI *registration* (so when calling the `POST /sc/ki` operation). So, if it is something users would want to change each time they activate a KI or whether it is something that you know when registering a KI and it does not change between activations. I am leaning towards the latter, so that would mean configuring it when registering a KI. This option might, however, be a bit more tedious to implement.

**Comment by JackJackie** [comment]
Hi @bnouwt,

Before diving into building the optional flag, I ran the unit tests of the smart connector.

There were 60 runs without failures and errors. 3 unit tests were skipped, but I saw that they were `@disabled`, so this makes sense.

File: smart-connector-rest-server/src/main/resources/openapi-sc.yaml
Comment (JackJackie) [reply smart-connector-rest-server/src/main/resources/openapi-sc.yaml:681]: All right...then let's go for the latter option...provide the option when registering a KI...I'll come up with a proposal for that.

Comment (JackJackie) [reply smart-connector-rest-server/src/main/resources/openapi-sc.yaml]: @bnouwt, I implemented the `knowledgeGapsEnabled` flag in the RestAPI, which runs through the nested Ask knowledge interaction execution of the knowledge engine. In addition, I made 4 tests to check the behavior of the SmartConnector java API. See: [here](https://github.com/TNO/knowledge-engine/tree/rest_api_with_knowledge_gaps/smart-connector/src/test/java/eu/knowledge/engine/smartconnector/api). Next step is to make some RestAPI tests to check the behavior at the RestAPI.

Comment (bnouwt) [reply smart-connector-rest-server/src/main/resources/openapi-sc.yaml]: Hey @JackJackie, thanks! That looks very thorough 👍  I will take a closer look the coming days.

Regarding the REST API tests, it would indeed be best to add those, but they are also quite some effort, so it might not be worth it.

File: smart-connector-rest-server/src/main/resources/openapi-sc.yaml
Comment (bnouwt) [inline smart-connector-rest-server/src/main/resources/openapi-sc.yaml:224]: Hey @JackJackie, just tuning in unasked 😄 . I like this description, but maybe move it from the example to the actual definition of the property? Not sure if it is possible, but similar to something like [this](https://github.com/TNO/knowledge-engine/blob/d482621eb38f7c0f243110c2ab55ca671444f02a/smart-connector-rest-server/src/main/resources/openapi-sc.yaml#L559)?

Comment (JackJackie) [reply smart-connector-rest-server/src/main/resources/openapi-sc.yaml:231]: I replaced the description to the head of the request body part.

**Comment by JackJackie** [comment]
Hi @bnouwt , I finished the unit tests at RestAPI level...if it is up to me, we are done with this merge request...can you please review one more time and see whether we can merge and resolve possible conflicts? Thx, Jack

File: smart-connector-rest-dist/src/test/java/eu/knowledge/engine/rest/api/TestAskAnswerReactWithGapsEnabled.java
Comment (bnouwt) [inline smart-connector-rest-dist/src/test/java/eu/knowledge/engine/rest/api/TestAskAnswerReactWithGapsEnabled.java:192]: Hey @JackJackie, all of these tests succeed individually and during a Maven build, but some of them fail when I run them directly from my Eclipse IDE environment. Apparently, in that scenario all these tests run after one another in the same JVM and thus using the same KE Runtime. And since all tests (also the older ones) do fully not clean up their KBs before finishing (they only stop the REST Server, but the KE Runtime stays active and so the registered KBs remain active), if the next test uses the same Knowledge Base IDs (which the above test does with for example `https://www.tno.nl/example/relationAsker`) these will already be registered and give an http status code 400 with the following error:

`{"messageType":"error","message":"That knowledge base ID is already in use."}`

I contributed to this branch and created a generalized solution for the above problem by introducing the `TestUtil#unregisterAllKBs(String)` class and method that remove all registered KBs from a REST Server automatically. I call this method at the end of every REST API Unit Test and this indeed seems to fix it.

File: reasoner/src/main/java/eu/knowledge/engine/reasoner/api/TriplePattern.java
Comment (bnouwt) [inline reasoner/src/main/java/eu/knowledge/engine/reasoner/api/TriplePattern.java:181]: remove this line

File: smart-connector-rest-dist/src/test/java/eu/knowledge/engine/rest/api/TestAskAnswerWithGapsEnabled.java
Comment (bnouwt) [inline smart-connector-rest-dist/src/test/java/eu/knowledge/engine/rest/api/TestAskAnswerWithGapsEnabled.java:126]: For me this assertion failed 8 out of 10 times when running all REST API Unit Tests from Eclipse IDE. This was caused by a timing issue where the `relationProvider` KB was not ready yet, but the `relationAsker` already activated its `AskKnowledgeInteraction`. Fixed it using a `CountDownLatch` to synchronize between these two different threads.

**Comment by bnouwt** [comment]
I've merged the `master` branch into this branch and resolved the conflicts. This was not trivial, because the `master` also contained changes to the `*KnowledgeInteraction` classes. I am currently running a maven build and hopefully this succeeds and we can let GitHub check if this branch is ready to be merged.

**Comment by bnouwt** [comment]
Unfortunately, the build did not yet succeed. The `DistributedMessageDispatcherTest#testRemoteMessageExchange` test somehow fails, but it might be a fluke. For now I've committed the changes and hopefully they build on github.

**Comment by bnouwt** [comment]
Hi @JackJackie, as mentioned above the build failed due to the REST API Knowledge Gap Unit Tests failing and I found that this is caused by the fact that the default `MatchStrategy` is not sufficient to find Knowledge Gaps. Unfortunately, it is not possible yet to change the default MatchStrategy from the REST API.

I see a few options to fix this:
- replace the smart connector REST API reasonerEnabled configuration with a matchStrategy configuration.
- use conditional configuration like `if knowledgeGapsEnabled, then MatchStrategy.SUPREME` to automatically upgrade the MatchStrategy if Knowledge Gaps are requested.
- add a REST API Knowledge Interaction configuration for matchStrategy.

The conditional configuration could make configuring the Knowledge Engine very complex, so I am not sure I would go for that one.

This reason why this problem did not pop-up before is because I merged the `master` branch into this branch including the new graph pattern matching algorithm with configuring match strategies at the Knowledge Interaction level.

File: smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/AskResult.java
Comment (bnouwt) [inline smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/AskResult.java:108]: @JackJackie Something to consider is to not only make knowledge gaps available on `AskResult`, but also on `AskPlan`. In some scenario's you want to determine knowledge gaps without actually executing the `AskKnowledgeInteraction`. The `admin-ui` could, for example, use this to show which knowledge interactions have knowledge gaps, but we do not want to have to execute the `AskKnowledgeInteraction` before determining the knowledge gaps.

Since the REST API does not distinguish between `planning` and `executing`, I think only the Java API should support this. Do you think we should include this to this Pull Request (PR), or do we just want to finally close it?

Comment (JackJackie) [reply smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/AskResult.java:108]: All right @bnouwt : let's discuss when you have the opportunity!

File: reasoner/src/main/java/eu/knowledge/engine/reasoner/api/TriplePattern.java
Comment (JackJackie) [reply reasoner/src/main/java/eu/knowledge/engine/reasoner/api/TriplePattern.java:187]: Please do this Barry!!

**Comment by JackJackie** [comment]
@bnouwt : I discovered a small bug in the ProactiveApiServiceImpl around the knowledgeGaps being null.....I need to fix that before we can continue

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (JackJackie) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:157]: So, I had to rethink what is going on here. We need to distinguish between the following situations:

1. The KI is not knowledgeGapsEnabled => no knowledge gaps will be returned
2. The KI is knowledgeGapsEnabled AND there are bindings found => knowledge gaps will be empty
3. The KI is knowledgeGapsEnabled AND there are no bindings found => knowledge gaps will contain the gaps found (by the way, might be empty as well).

The above code does not cover situation 2 well as there was a null result for the knowledge gaps instead of an empty one.

Comment (bnouwt) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:157]: Hi @JackJackie, as discussed yesterday I agree that the above situations are indeed how we would like the REST API te respond 👍 

Were you already able to implement this? Currently, the `TestAskWithGapsNotEnabled#testAskWithoutGaps` unit test fails because it returns knowledge gaps even when they are not enabled:

```
Error: [ERROR] eu.knowledge.engine.rest.api.TestAskWithGapsNotEnabled.testAskWithoutGaps -- Time elapsed: 1.468 s <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <false> but was: <true>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertFalse.failNotFalse(AssertFalse.java:63)
	at org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:36)
	at org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:31)
	at org.junit.jupiter.api.Assertions.assertFalse(Assertions.java:231)
	at eu.knowledge.engine.rest.api.TestAskWithGapsNotEnabled.testAskWithoutGaps(TestAskWithGapsNotEnabled.java:64)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

Comment (bnouwt) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:157]: I've fixed one of the REST API unit tests by making sure the ordering of triples within a knowledge gap are not important for the assertion. It failed for me, because the ordering was different.

File: smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/AskResult.java
Comment (bnouwt) [reply smart-connector-api/src/main/java/eu/knowledge/engine/smartconnector/api/AskResult.java:108]: Decided to add this as a separate issue: https://github.com/TNO/knowledge-engine/issues/561

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (JackJackie) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:157]: Yes, I noticed that as well....now I'm dealing with the fact that the AskResult in the RestAPI needs to give back 2 different objects as part of the '200' response....one with knowledgeGaps and one without....to be continued!!

**Comment by JackJackie** [comment]
@bnouwt : above here it says...."Changes requester...1 change requested"...I tried to look which change request that is, but couldn't find out. Do you know? PS: let's not merge yet, I need to deal with the issue of having 2 possible '200' responses first.

File: smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java
Comment (JackJackie) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:158]: Did take care of it...check it out!

Comment (bnouwt) [reply smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java:158]: Good to see you got it sorted out 👍 , although having an additional `AsjResultWithGaps` seems a bit overkill. Maybe we could also try this solution? 

https://stackoverflow.com/a/72734037

Not sure if it works and it might break other things, but configuring it seems like a 'cleaner' solution. What do you think?

**Comment by JackJackie** [comment]
@bnouwt , @Sophietje , I'm merging the RESTapi with knowledge gaps enabled branch into main....so you can continue with other features based on this new version....hopefully nothing (much) breaks :-)

**Pull request merged**
